### PR TITLE
Convert all drawings to freezed data classes

### PR DESCRIPTION
**Pull request by roscale**
All drawings have been converted to `freezed` data classes. No advanced features of the package have been used yet. The `@immutable` annotations have been repalced with `@freezed` and the code has been rearranged to how `freezed` expects it. `operator=`, `hashCode`, and `toString` methods have been deleted because `freezed` generates them for us.

`GYWDrawing` is converted into an interface to make it work, but the subtyping relation between concrete drawings and `GYWDrawing` stays the same.

The only API change is that the icon field in `IconDrawing` is now named.

### CODE REVIEW

**Comment by roscale** [comment]
You can ignore `drawings.freezed.dart` for obvious reasons. It's generated code that you're not supposed to review.

**Comment by roscale** [comment]
Am I supposed to modify the changelog? This is one step out of n (probably 2 or 3) to properly integrate freezed. Are we adding one entry in the changelog for each step?

**Comment by TheToto42** [comment]
> Am I supposed to modify the changelog? This is one step out of n (probably 2 or 3) to properly integrate freezed. Are we adding one entry in the changelog for each step?

Yes, you should. Specify in which way `freezed` is integrated. I usually squash some changes in the deployment commit.

**Comment by TheToto42** [comment]
> You can ignore `drawings.freezed.dart` for obvious reasons. It's generated code that you're not supposed to review.

Shouldn't we put that in the `.gitignore` file then ?

File: lib/src/drawings.dart [hunk 1]
```diff
@@ -2,27 +2,21 @@ import "dart:convert";
 import "dart:typed_data";
 
 import "package:flutter/material.dart";
+import "package:flutter_gyw/flutter_gyw.dart";
```
Comment (TheToto42) [inline lib/src/drawings.dart:5]: Inside src folder, I prefer to call the src file you need instead of the entire library to prevent cycle import.

File: test/drawing_test.dart [hunk 2]
```diff
@@ -70,10 +70,10 @@ void main() {
       final GYWIcon icon = GYWIcons.left.icon;
 
       final GYWDrawing drawing = IconDrawing(
-        icon,
+        icon: icon,
         left: 120,
         top: 220,
-        color: Colors.green,
+        color: Colors.green.shade500,
```
Comment (TheToto42) [inline test/drawing_test.dart:76]: Why did you change the color ? You don't like the basic green ? :laughing:

**Review by TheToto42 — CHANGES_REQUESTED** [review]

**Comment by roscale** [comment]
> > You can ignore `drawings.freezed.dart` for obvious reasons. It's generated code that you're not supposed to review.
> 
> Shouldn't we put that in the `.gitignore` file then ?

I don't think you can. Users that depend on our library must have all the generated files to be able to use our classes. In application code, I would have said ok because you can regenerate them on the fly after you checkout, but flutter doesn't touch third party files when you import a dependency.

If you look at this library, you can see that they too are committing these files:
https://github.com/babakcode/flutter_gemini/tree/master/lib/src/models/gemini_model

File: test/drawing_test.dart (hunk 2 above)
Comment (roscale) [reply test/drawing_test.dart:76]: Actually, `Colors.green` is a family of colors, and its type is `MaterialColor`. `Colors.green.shade500` is a particular Color. The test was failing on my machine because of this.

Comment (roscale) [reply test/drawing_test.dart:76]: Oh, I know why it fails now and didn't fail before. Apparently, this is false:
`expect(Colors.green, Color(Colors.green.value));`

Before, we had our own equality operator on IconDrawings that compared color.value, not color directly, so we were bypassing the equality operator of Color.

Freezed just compares the color instances, and a MaterialColor is not a Color.

Comment (roscale) [reply test/drawing_test.dart:76]: I don't think it's necessarily a problem imo. I don't think anybody will realistically serialize, deserialize, and then compare objects like we do in tests.

**Comment by TheToto42** [comment]
> I don't think you can. Users that depend on our library must have all the generated files to be able to use our classes. In application code, I would have said ok because you can regenerate them on the fly after you checkout, but flutter doesn't touch third party files when you import a dependency.

This seems correct. Can you look for a Github actions that checks that the generated freezed code is the one that is up-to-date? You can inspire yourself with how golden tests are checked.

File: test/drawing_test.dart (hunk 2 above)
Comment (TheToto42) [reply test/drawing_test.dart:76]: At first, I would agree with you, but after a bit of thinking, shouldn't we store the color value in hex and have a getter with the material color then ? I'm scared of users using the equality operator and not knowing that.

Comment (roscale) [reply test/drawing_test.dart:76]: Sure, we can do that. It's true that nobody will know the reason the equality fails.

**Comment by roscale** [comment]
> This seems correct. Can you look for a Github actions that checks that the generated freezed code is the one that is up-to-date? You can inspire yourself with how golden tests are checked.

I didn't find any GitHub action that checks that. However, in my experience, when the generated code is out of date, the code simply won't compile, so the linter will catch the error.

**Review by TheToto42 — APPROVED** [review]

File: test/drawing_test.dart [hunk 3]
```diff
@@ -70,10 +70,10 @@ void main() {
       final GYWIcon icon = GYWIcons.left.icon;
 
       final GYWDrawing drawing = IconDrawing(
-        icon,
+        icon: icon,
         left: 120,
         top: 220,
-        color: Colors.green,
+        colorHex: Colors.green.value,
```
Comment (TheToto42) [inline test/drawing_test.dart:76]: Even if we store the value itself, can't we have the constructor that takes a Color object, it will be easier to use I think ?

Comment (roscale) [reply test/drawing_test.dart:76]: I have scouted the interwebs and I don't think that's possible. Freezed models are supposed to be PODs (plain old data structures) without OOP features like encapsulation. Otherwise, copyWith would also need to change to avoid leaking private API.

**Comment by roscale** [comment]
This PR is closed because we are replacing freezed with dart_mappable, which is less magical and more flexible.
